facebook/translation-differ.R

```diff
@@ -0,0 +1,240 @@
+##
+## Writes the lists of new and changed items to STDOUT, so redirect STDOUT to
+## your desired location.
+
+
+#### Background on .qsf files
+
+## A .qsf file is a json containing two elements, SurveyEntry with survey
+## metadata (start date, ID, name, etc) and SurveyElements with a numbered list
+## of survey components. For the purpose of finding differences in survey
+## contents, we are only interested in examining the SurveyElements item.
+
+## SurveyElement types are BL (block), FL (flow), SO (survey options), SCO
+## (scoring), PROJ (?), notes, STAT (statistics), QC (question count), SQ
+## (survey questions), and RS (response set). Detailed info:
+## https://gist.github.com/ctesta01/d4255959dace01431fb90618d1e8c241
+
+## We are only interested in block information ("BL") and survey questions
+## ("SQ"). The single BL item lists questions that appear in each block. The
+## "Trash" block contains not-displayed items. Within each SQ item, details are
+## stored in the Payload element, which can contain the following fields:
+
+# [1] "QuestionText"         "QuestionType"         "Selector"             "Configuration"        "QuestionDescription"  "Choices"              "ChoiceOrder"         
+# [8] "Validation"           "AnalyzeChoices"       "Language"             "QuestionText_Unsafe"  "DataExportTag"        "QuestionID"           "DataVisibility"      
+# [15] "NextChoiceId"         "NextAnswerId"         "DefaultChoices"       "SubSelector"          "DisplayLogic"         "GradingData"          "Answers"             
+# [22] "AnswerOrder"          "ChoiceDataExportTags" "Randomization"        "RecodeValues"         "DynamicChoices"       "DynamicChoicesData"   "SearchSource"        
+# [29] "QuestionJS"  
+
+## The meaning of "Answers" and "Choices" differs for matrix vs non-matrix
+## items. "Choices" list the vertical components -- subquestions for matrix
+## items and answer choices for non-matrix items. "Answers" list the answer
+## choices for matrix items and are missing for non-matrix items.
+
+
+suppressPackageStartupMessages({
+  library(jsonlite)
+  library(stringr)
+})
+
+#' Load and diff chosen translation files.
+#'
+#' @param old_qsf_path path to older Qualtrics translation file in .qsf format
+#' @param new_qsf_path path to newer Qualtrics translation file in .qsf format
+diff_qsf_files <- function(old_qsf_path, new_qsf_path) {
+  old_qsf <- get_qsf_file(old_qsf_path)
+  new_qsf <- get_qsf_file(new_qsf_path)
+  
+  diff_surveys(old_qsf, new_qsf)
+  
+  return(NULL)
+}
+
+#' Fetch and format a single .qsf translation file, keeping block and question info
+#'
+#' @param path path to Qualtrics translation file in .qsf format
+#'
+#' @return A named list
+get_qsf_file <- function(path) {
+  # Read file as json.
+  qsf <- read_json(path)
+  
+  ## Block
+  block_filter <- sapply(qsf$SurveyElements, function(elem) { elem[["Element"]] == "BL" })
+  block_out <- qsf$SurveyElements[block_filter][[1]]$Payload
```

Comment:
Seems like a use case for `Filter()`?

Reply:
The object is a list of nested lists, not a dataframe, so 'filter' isn't applicable here. I was originally trying to manipulate the qsfs as dataframes, but changes in fields between questions/waves means it is sometimes hard to convert to dataframe, so I'm handling the qsfs exclusively as lists.

Reply:
Right, but I meant capital-F `Filter()`, not `filter()`. But maybe `Filter()` is for vectors and not for lists, I'm not sure